JavaScriptCodeConventions.md

```diff
@@ -0,0 +1,58 @@
+# JavaScript Code Conventions
+> NOTE: These conventions also apply to super-sets of JavaScript which are used within Ultimaker such as TypeScript.
+
+## Introduction
+These JavaScript style guidelines are intended to be used by both the R&D and marketing team as part of a broader effort to standardize JavaScript style across the company and as an early facilitator for the fact that Ultimaker continues to offer more web-based products such as the Cura Cloud services and Cura Connect.
+
+## AirBnB as Baseline Style
+To reduce the need to build comprehensive style guidelines ourselves, we have decided to adopt the well-known and widely used **AirBnB style**. The full documentation for AirBnB style [can be found on GitHub](https://github.com/airbnb/javascript). For the most part, this style is already very similar to what is used by front-end teams at Ultimaker with a few differences, which are outlined below. It was also chosen because of it was designed for usage with React.js, which is used by the R&D department. It is applicable to Vue.js as well, however, and can be used by the marketing department.
+
+## Enforcing AirBnB Style in Your Editor
+### Set-Up for VS Code
+1. Install packages globally:
+
+```
+npm i -g eslint eslint-config-airbnb eslint-plugin-import eslint-plugin-jsx-a11y eslint-plugin-react
+```
+2. Install the [`eslint` plugin for VS Code](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint).
+3. Create an `.eslintrc` file in your project’s root directory.
+4. Add the following content to `.eslintrc`:
+
+```
+{
+  "extends": "airbnb",
+  "env": {
+    "node": true,
+    "es6": true,
+    "browser": true
+  },
+  "rules": {
+	  no-underscore-dangle: "allow"
+  }
+}
+```
+
+
+## Required Changes for Adoption
+### Simple to Change
+Some changes can easily be made with a single find & replace operation and a single commit.
+  
+#### Tabs
+- **Existing:** Use 4 spaces for tabs.
+- **AirBnB:** Use 2 spaces for tabs.
+
+#### Strings
+- **Existing:** Use single-quotes.
+- **AirBnB:** Use double-quotes.
+
+### Difficult to Change
+One change is very difficult to change and will need to be adapted over time. For this rule, an exception can be made until projects are converted to TypeScript, at which point the AirBnB style will be followed with an additional caveat.
+
+- **Existing:** Private properties/methods should start with an underscore (_).
+- **AirBnB:** JavaScript does not have a concept of private, so do not make “fake private” properties/methods with an underscore.
```

Comment:
Right. This is why there's an exception in the `.eslintrc` example to not enforce this rule so that it doesn't require big changes to vanilla JavaScript. But in the front-end guild meeting we agreed that we'd like to eventually use only TypeScript and as such, when a project gets migrated, this change should be part of the migration. For new projects started with TypeScript from this point forward, the `private` prefix should be utilized.

Reply:
Python doens't have a real private. But if you do use double underscore then Python will do name-mangling to make it unique to the class it is in. It is a bit of a hack but it is slightly more private than just a convention.

Reply:
Also TypeScript has a private keyword and it is enforced at compile time, but it has no effect at runtime. Classes/subclasses which define the same named private field can still collide (but the compile detects this).

There is a TC39 proposal for adding real private fields and variables to JS. It adds more syntax though and it is not pretty syntax at that.

Reply:
The single underscore is protected, not private. I'm not sure how much it really matters (I have no idea how OO the code is going to get)

Reply:
All of our React is very OO

Reply:
In that case; the distinction between private and protected is pretty darn important ;)